### ISSUE TITLE
Is azure-sdk-for-cpp Y2k38 safe?

### ISSUE THREAD

**Issue by bzhou-sw** [Azure/azure-sdk-for-cpp#5841]
**Query/Question**
Is `azure-sdk-for-cpp` safe for Y2k38 bug.

***Why is this not a Bug or a feature Request?***
Just want to check if dev team has done any code assessment to make sure timestamp won't be affected by Y2k38 bug.

**Setup (please complete the following information if applicable):**
 - OS: Linux
 - IDE : None
 - Version of the Library used:  azure-storage-files-shares_12.5.0-beta.1
 
 **Information Checklist**
 Kindly make sure that you have added all the following information above and checkoff the required fields otherwise we will treat the issuer as an incomplete report
- [ ] Query Added
- [ ] Setup information Added

**Comment by antkmsft** [comment]
I am sorry for long and detailed reply, I just want to be accurate and want to give a good explanation, so your expectations are correct.
If anything is not clear, you want a simpler answer for more specific question, please feel free to reopen and ask more questions until you clearly get an answer of what you were looking for.

Azure SDK internally stores DateTimes as 64-bit signed integers, one unit being equal to 100 nanoseconds (10^-7 seconds), from the start of "year 0". That is sufficient for the dates up to over year 29000. Practically, current Azure::DateTime::ToString() implementation limits that by the date range between year 0000 to year 9999 inclusive.

When Azure SDK talks to Azure services, and either sends or receives DateTime, it may send/parse it in one of 2 ways: as a string (either in RFC1123 format, or in RFC3339 format), or as a Unix timestamp (number of seconds since 1970).

Both string formats are good for up to at least year 9999.

Unix timestamps, we also convert to/from 64-bit singned integers, which is good for up to 292.5 billion years (but often that would be limited by Azure::DateTime's storage capacity, which can store, without converting to strings, datetimes of up to over a year 29000).

Now, that's as far as the SDK itself goes. The SDK talks to Azure services. And whether the Azure Service would correctly handle a date past year 2038 - it, in theory, could depend, and it could be different from service to service, and maybe even within the same service, region to region. Note that we as an SDK team do not control and we may not know what the Azure Service implementation does.

I think there is good probability that most of the Azure Services are Y2k38 safe even today. And I think there is very high probability all of them will be Y2k38 safe by year 2038.

So, when we talk to an Azure Service, we use certain API version, for example "2024-05-02". Hypothetically, the worst case scenario is that the service would not be supporting 2038 dates when the API version "2024-05-02" is used, but, let's say, it will be able to accept datetimes past 2038 when we communicate with the future API version of "2027-09-23", and they won't backport that to older API versions.
In that case, the program that you compiled in 2024, will not be able to work after 2038.
But if you take your source code, and compile it against the version of a client library that gets released in 2027 (so it uses an API version of "2027-09-23" for that service), it will be y2k38 safe. Most likely, you will be able to just rebuild it without modifying a single line of your code, given our approach to avoid/minimize breaking changes.

The best case scenario, is that you build your app today, and it works past year 2038 without recompilation. All would depend on whether the specific Azure service that you're using is Y2k38-ready.

To summarize, taking your question at face value, the "Azure SDK" itself does not have any problems with dates past Y2K38.

**Comment by bzhou-sw** [comment]
@antkmsft Thank you so much for your effort and time investigating this.
